packages/gator-permissions-snap/src/clients/accountApiClient.ts

```diff
@@ -2,37 +2,75 @@ import { logger } from '@metamask/7715-permissions-shared/utils';
 import { type Hex } from '@metamask/delegation-core';
 import {
   InvalidInputError,
+  LimitExceededError,
+  ParseError,
   ResourceNotFoundError,
   ResourceUnavailableError,
 } from '@metamask/snaps-sdk';
+import { z } from 'zod';
 
 import { ZERO_ADDRESS } from '../constants';
 import type { TokenBalanceAndMetadata } from './types';
 
+/**
+ * Zod schema for validating token balance response
+ */
+const TokenBalanceResponseSchema = z.object({
+  name: z.string().min(1).max(100),
+  symbol: z
+    .string()
+    .min(1)
+    .max(20)
+    .regex(/^[A-Z0-9]+$/iu, 'Symbol must contain only alphanumeric characters'),
+  decimals: z.number().int().min(0).max(77), // ERC-20 standard allows up to 77 decimals
+  type: z.string().min(1).max(20).optional(),
+  iconUrl: z.string().url().max(2048).optional(), // Limit URL length and validate format
+  coingeckoId: z.string().min(1).max(100),
+  address: z
+    .string()
+    .regex(/^0x[a-fA-F0-9]{40}$/u, 'Invalid hex address format'),
+  occurrences: z.number().int().min(0),
+  sources: z.array(z.string().min(1).max(100)).max(50), // Limit number of sources
+  chainId: z.number().int().positive(),
+  blockNumber: z
+    .string()
+    .regex(
+      /^(latest|\d+)$/u,
+      'Block number must be "latest" or numeric string',
+    ),
+  updatedAt: z.string().datetime(),
+  value: z.record(z.unknown()),
+  price: z.number().finite().min(0),
+  accounts: z
+    .array(
+      z.object({
+        accountAddress: z
+          .string()
+          .regex(/^0x[a-fA-F0-9]{40}$/u, 'Invalid hex address format'),
```

Comment:
nit: we repeat the address validation a couple of times. This could be decomposed into a separate zod definition `const zAddress = z.string().regex(...)`

edit: Infact, we already define this in `packages/shared/src/types/common.ts`: 

```
export const zAddress = z
  .string()
  .regex(/^0x[a-fA-F0-9]{40}$/u, 'Invalid Ethereum address')
  .transform((value) => {
    return getChecksumAddress(value as Hex);
  });
```